File: backend/app/api/management.py
"""Management 面板聚合接口：返回所有非 admin 智能体的摘要信息。"""

from fastapi import APIRouter
from loguru import logger

from app.core.config import AGENTS_CONFIG
from app.core.auth import _user_agents_var
from app.service.sessions_service import load_sessions
from app.service.messages_service import load_messages

router = APIRouter(prefix="/api/management", tags=["management"])

# 供应商 → 默认模型映射，与 app.providers.factory._default_model_for 保持同步
_DEFAULT_MODELS: dict[str, str] = {
    "deepseek": "deepseek-chat",
    "openai": "gpt-4o",
    "moonshot": "kimi-k2.5",
}


def _resolve_model(cfg: dict) -> str:
    """从 agent 配置解析显示用模型名。

    优先级：
      1. cfg.model（YAML 中显式指定）
      2. cfg.provider → 已知映射
      3. 兜底 "deepseek-chat"
    """
    explicit = cfg.get("model")
    if explicit:
        return explicit
    provider = cfg.get("provider", "deepseek")
    return _DEFAULT_MODELS.get(provider, f"{provider}-chat")


def _build_agent_summary(agent_id: str, cfg: dict) -> dict:
    """为单个 agent 构建摘要（会话数、最近回复时间、最近会话标题）。"""
    name = cfg.get("name", agent_id)
    locked = cfg.get("locked", False)

    sessions = load_sessions(agent_id)
    session_count = len(sessions)
    last_session_title = sessions[0]["title"] if sessions else None

    last_reply_time = None
    if sessions:
        first_session_id = sessions[0]["session_id"]
        messages = load_messages(agent_id, first_session_id)
        # 倒序查找最后一条 assistant 消息的 created_at
        for msg in reversed(messages):
            if msg.get("role") == "assistant":
                last_reply_time = msg.get("created_at")
                break

    model = _resolve_model(cfg)

    return {
        "id": agent_id,
        "name": name,
        "locked": locked,
        "model": model,
        "session_count": session_count,
        "last_reply_time": last_reply_time,
        "last_session_title": last_session_title,
    }


@router.get("/agents-summary")
async def get_agents_summary():
    """返回所有非 admin 智能体的聚合摘要列表。"""
    agents = []

    # 系统 agent（config/agents/*.yaml + config.yaml agents）
    for agent_id, cfg in AGENTS_CONFIG.items():
        if agent_id == "admin":
            continue
        if not isinstance(cfg, dict):
            continue
        agents.append(_build_agent_summary(agent_id, cfg))

    # 当前用户的 custom agent
    try:
        user_agents = _user_agents_var.get()
        for agent_id, cfg in user_agents.items():
            if agent_id in AGENTS_CONFIG:
                continue  # 已在系统 agent 中
            agents.append(_build_agent_summary(agent_id, cfg))
    except LookupError:
        pass

    logger.info("management agents-summary: count={}", len(agents))
    return {"agents": agents}

File: backend-rs/src/routes/management.rs
use axum::extract::Extension;
use axum::routing::get;
use axum::{Json, Router};
use serde::Serialize;
use serde_json::{json, Value};

use crate::core::auth::UserContext;
use crate::core::config::AgentConfig;
use crate::provider::factory;

#[derive(Serialize)]
struct AgentSummary {
    name: String,
    title: String,
    locked: bool,
    model: String,
    session_count: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    last_reply_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    last_session_title: Option<String>,
}

pub fn router() -> Router {
    Router::new().route("/api/management/agents-summary", get(get_agents_summary))
}

async fn get_agents_summary(Extension(ctx): Extension<UserContext>) -> Json<Value> {
    let config = crate::core::config::get_config();
    let mut agents: Vec<AgentSummary> = Vec::new();

    // 系统 agent（config/agents/*.yaml + config.yaml agents）
    for (agent_id, cfg) in &config.agents {
        if agent_id == "admin" {
            continue;
        }

        let sessions = crate::service::sessions::load_sessions(agent_id);
        let session_count = sessions.len();
        let last_session_title = sessions.first().map(|s| s.title.clone());

        let last_reply_time = sessions.first().and_then(|s| {
            let messages = crate::service::messages::load_messages(agent_id, &s.session_id);
            messages
                .iter()
                .rev()
                .find_map(|m| {
                    if m.role == "assistant" {
                        Some(m.created_at.clone())
                    } else {
                        None
                    }
                })
        });

        let model = resolve_model(cfg);

        agents.push(AgentSummary {
            name: agent_id.clone(),
            title: cfg
                .title
                .clone()
                .unwrap_or_else(|| crate::core::config::DEFAULT_AGENT_TITLE.to_string()),
            locked: cfg.locked.unwrap_or(false),
            model,
            session_count,
            last_reply_time,
            last_session_title,
        });
    }

    // 当前用户的 custom agent（DATA_DIR/u_{user_id}/agent/*.yaml）
    for (agent_id, cfg) in &ctx.user_agents {
        if agent_id == "admin" {
            continue;
        }
        if config.agents.contains_key(agent_id) {
            continue; // 已在系统 agent 中，跳过
        }

        let sessions = crate::service::sessions::load_sessions(agent_id);
        let session_count = sessions.len();
        let last_session_title = sessions.first().map(|s| s.title.clone());

        let last_reply_time = sessions.first().and_then(|s| {
            let messages = crate::service::messages::load_messages(agent_id, &s.session_id);
            messages
                .iter()
                .rev()
                .find_map(|m| {
                    if m.role == "assistant" {
                        Some(m.created_at.clone())
                    } else {
                        None
                    }
                })
        });

        let model = resolve_model_from_value(cfg);

        agents.push(AgentSummary {
            name: agent_id.clone(),
            title: cfg
                .get("title")
                .and_then(|v| v.as_str())
                .unwrap_or(crate::core::config::DEFAULT_AGENT_TITLE)
                .to_string(),
            locked: cfg.get("locked").and_then(|v| v.as_bool()).unwrap_or(false),
            model,
            session_count,
            last_reply_time,
            last_session_title,
        });
    }

    Json(json!({"agents": agents}))
}

/// 从 config 解析显示用模型名
fn resolve_model(cfg: &AgentConfig) -> String {
    if let Some(model) = cfg.extra.get("model").and_then(|v| v.as_str()) {
        return model.to_string();
    }
    let provider = cfg
        .extra
        .get("provider")
        .and_then(|v| v.as_str())
        .unwrap_or("deepseek");
    factory::default_model_for(provider).to_string()
}

/// 从 Value 类型的配置解析显示用模型名（供 custom agent 使用）
fn resolve_model_from_value(cfg: &Value) -> String {
    if let Some(model) = cfg.get("model").and_then(|v| v.as_str()) {
        return model.to_string();
    }
    let provider = cfg
        .get("provider")
        .and_then(|v| v.as_str())
        .unwrap_or("deepseek");
    factory::default_model_for(provider).to_string()
}
